python/ray/_raylet.pyx

```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
     return TaskID(object_id.native().TaskId().Binary())
 
 
+def setup_logging(stdout_name, stderr_name):
+    """Sets up logging for the current worker, creating the file and flushing
+    buffers as is necessary.
+
+    Args:
+        stdout_name (str): The file name that stdout should be written to.
+        stderr_name(str): The file name that stderr should be written to.
+
+    Returns:
+        (tuple) The absolute paths of the files that stdout and stderr will be
+    written to.
+    """
+    stdout_path = ""
+    stderr_path = ""
+    worker_pid = os.getpid()
+
+    cdef extern from "ray/util/util.h":
+        void flush_out()
+        void flush_err()
+
+    if stdout_name:
+        # Line-buffer the output (mode 1).
+        stdout_file = open_worker_log(stdout_name, worker_pid)
+
+        # Before redirecting stdout we need to make sure userspace buffers
+        # are all flushed.
+        # Flush python's userspace buffers
+        sys.stdout.flush()
+
+        # TODO (Alex): Flush the c/c++ userspasce buffers if necessary.
+        # `fflush(stdout); cout.flush();`
+
+        stdout_fileno = sys.stdout.fileno()
```

Comment:
Here's the thing:

(1) This is the existing code behavior (and there doesn't seem to be a problem).
(2) We control how workers are started. It would be quite hard to start a worker without calling `ray start`.
(3) The underlying issue is that IDLE has a poorly design stdout object (i.e. it doesn't follow the Liskov Substitution Principle). 

I don't think need to (or even should) introduce additional code to support bad practices, which no one uses, in order to make the failure messages a little nicer (either way we're going to fail on startup anyways).

Reply:
I actually don't think it's an LSP violation; it's consistent with the file protocol, and I'm not aware of any extra requirements on `stdout` et al. :-) [`IOBase`](https://docs.python.org/3/library/io.html#io.IOBase.fileno) does say this:

> Return the underlying file descriptor (an integer) of the stream if it exists. An `OSError` is raised if the IO object does not use a file descriptor.

I'm not really worried about whether it's bad practice or not (I can argue both ways), but I'm worried about the UX. Programs often have to work around environment idiosyncrasies to provide a better experience.

As to why this hasn't come up yet, I'm unclear on whether this also applies to the driver or not. If it does, and it hasn't come up, my guess is it's most likely due to the fact that IDLE tends to be used on Windows, which we didn't previously support. I think Linux/Mac users rarely run IDLE and instead type `python` directly in their terminals.

That said, if you feel strongly about it regardless, I'll probably just end up changing it myself later, so it's not a big deal either way. Just wanted to point it out because having various packages break in IDLE due to this issue is something I encounter fairly regularly on Windows.

Reply:
This should be worker only. I don't have a windows machine (or even IDLE) readily available for testing, but if you can currently run `ray.init()` in IDLE then this won't change that functionality. If you REALLY want I can add a try/catch, but I believe the correct behavior should be to fail and propagate an error message up if someone tries to start a worker in IDLE.

Reply:
Ah ok, I guess feel free to keep it as is then. Maybe it won't come up. I can fix it if it comes up haha.